scaaml/capture/input_generators/attack_point_iterator_models.py

```diff
@@ -165,3 +166,177 @@ def items(self) -> AttackPointIteratorT:
+                pass
+        if not self.operands:
+            return 0
+        else:
+            # If `non_negative_lengths` is empty it means that all
+            # operands are infinite.
+            smallest_length: int = min(non_negative_lengths, default=-1)
+            if smallest_length < 0:
+                raise LengthIsInfiniteException
+            return smallest_length
+
+    def items(self) -> AttackPointIteratorT:
+        items_of_operands: List[AttackPointIteratorT] = []
+        for operand in self.operands:
+            items_of_operands.append(operand.items())
+        return iter(
+            self._merge_dictionaries(tuple_of_dictionaries)
+            for tuple_of_dictionaries in zip(*items_of_operands))
+
+    @staticmethod
+    def _merge_dictionaries(
+        tuple_of_dictionaries: Tuple[Dict[str,
+                                          List[int]]]) -> Dict[str, List[int]]:
+        merged_dictionary = {}
+        for value in tuple_of_dictionaries:
+            merged_dictionary.update(value)
+        return merged_dictionary
+
+    def get_generated_keys(self) -> List[str]:
+        generated_keys = []
+        for operand in self.operands:
+            generated_keys += operand.get_generated_keys()
+        return generated_keys
+
+
+class CartesianProductIteratorModel(BaseModel):
+    """
+    Attack point iterator cartesian product pydantic model. This class takes
+    any amount of operands and combines them just like a cartesian product 
+    would.
+
+    Args:
+        operation (Literal['cartesian_product']): The operation of the iterator
+            represents what the iterator does and what 
+            has to be in the config file. This is only used once to
+            double check if the operation is the correct one.
+            
+        operands (List[Union[BasicIteratorModels, RepeatIteratorModel]]):
+           The operands are any number of BasicIteratorModels or
+            RepeatIteratorModels that will be combined. If the operands list
+            is empty it will raise a ValueError. If one of the operands
+            length is 0 the length of the cartesian product iterator will
+            also be 0, it will return an empty iterator. If one of the
+            operands iterates infinitely it will throw a
+            LengthIsInfiniteException in the init.
+    """
+    operation: Literal["cartesian_product"]
+    operands: List[Union[BasicIteratorModels, RepeatIteratorModel,
+                         "ComplicatedIteratorModel"]]
+
+    @model_validator(mode="after")
+    def check_model(self) -> "CartesianProductIteratorModel":
+        if len(self.operands) > 1:
+            self.operands = [
+                self.operands[0],
+                CartesianProductIteratorModel(operation="cartesian_product",
+                                              operands=self.operands[1:])
+            ]
+        else:
+            self.operands = [self.operands[0]]
```

Comment:
I do not really understand this line. Could we have a documentation for it? I am not even sure it does what it is supposed to do.

Reply:
it does what it is supposed to do. We have the same code in the original attack point iterators and I tested this one to see if it works.

Reply:
I guess this line doesn't really do anything

Reply:
Well it does. It raises an exception when `self.operands` is an empty list. Which is not very user-friendly (kind of unexpected exception somewhere deep in the code, no descriptive message).

Reply:
Does it because I thought pydantic makes you input something or does it allow empty lists []?